### DISCUSSION TITLE
Define variable inside template script to be used later in the same script.

### DISCUSSION THREAD

**Discussion by waltergallegog** [nextflow-io/nextflow#4008] [Q&A]
Hello,

I'm using the following template:
```bash
sv_ref_gz="${sv_ref}.gz"
bgzip -c $sv_ref > \$sv_ref_gz
tabix -f -p vcf \$sv_ref_gz

some_python_script.py --sv \$sv_ref_gz ...
```

which is called from a process:
```groovy
process svValCute {
    input:
        path sv_ref
    output:
        path '*vcf'
    script:
        template 'sv_val.sh'
}
```
In the template, in the first line, I'm defining the variable `sv_ref_gz`. I want to reference this new variable in the next 3 commands.

The template works when referencing the variable with `\$` , as it is a bash variable. But according to the documentation at https://www.nextflow.io/docs/latest/process.html#template, this is not a good practice, because the `\$` will not be evaluated correctly when running from a terminal.

Is there a way I can use this variable so that:
 - It can be interpreted by Nextflow when running the template from a process.
 - It can be interpreted by bash when running the template directly from a terminal.

Using `$sv_ref_gz` gives me the following error when running from nextflow:
```
No such property `sv_ref_gz` -- check template file
```

Thanks for the support,
Walter

**Comment by manuelesimi** [comment]
Hello,

Can you try the following?

```groovy
process svValCute {
    input:
        path sv_ref
    output:
        path '*vcf'
    shell:
        template 'sv_val.sh'
}

```

and then:
``` bash
sv_ref_gz="!{sv_ref}.gz"
bgzip -c $sv_ref > $sv_ref_gz
tabix -f -p vcf $sv_ref_gz

some_python_script.py --sv $sv_ref_gz ...
```

**Reply by waltergallegog** [reply]
Hello @manuelesimi 
Thanks for the suggestion, but the `!{var}` syntax is not understood by bash, so it would not be possible to run the template you propose directly from the terminal.
BR
Walter.

**Reply by manuelesimi** [reply]
I'm not sure how you can obtain what you describe in the script. Nextflow needs to distinguish between vars coming from the channels and bash vars. The solution I was proposing minimizes the impact on the script, since you don't have to escape all the bash variables, but only the (few) coming from the channels.
